File: cs415-Parallel-Programming/cpp-java-consumer-producer/prodcons1.cpp
/*
    Brooke Czerwinski
    CS 415 Parallel Programming
    Assignment 1

    prodcons1.cpp

    C++ program to use parallel thread to add to and remove from a queue.

    References used:
    https://en.cppreference.com/w/cpp/thread/condition_variable
    https://man7.org/linux/man-pages/man3/sched_getcpu.3.html
    https://www.cplusplus.com/reference/condition_variable/condition_variable/
    https://www.cplusplus.com/reference/condition_variable/condition_variable/wait/
    https://www.baeldung.com/java-static
*/

#include "queue.h"
#include <iostream>
#include <thread>
#include <mutex>
#include <condition_variable>
#include <sched.h>

using std::cin;
using std::cout;
using std::endl;
using std::mutex;
using std::thread;
using std::condition_variable;
using std::unique_lock;



// Required Global Parameters
int     BUFSIZE = 20;       // queue capacity
int     NUMITEMS = 100;     // total number of data items

// Other Global Parameters
Queue               *aQueue;    // Queue pointers
mutex               mtx;          // Mutex variable
mutex               iosMtx;
condition_variable  cv;         // Condition Variable
bool                addToQueue = true;
bool                removeFromQueue = false;

bool                ioReady = false;


// Function prototypes
void producer();
void consumer();

int main()
{
    // Initialize queue
    aQueue = new Queue(BUFSIZE);

    // Create threads
    thread  producerThread(producer);
    thread  consumerThread(consumer);

    
    ioReady = true;
    cv.notify_one();


    // Join threads
    producerThread.join();
    consumerThread.join();

    // Final message
    cout << "Main: all done!" << endl;
    
    // Release memory
    delete aQueue;

    return 0;
}

void producer()
{
    // Uses curly braces to establish a new scope so that "lck" will get 
    // destroyed (and lock released) when it goes out of scope.
    {
        // Obtain a lock on the i/o so messages don't get mixed together
        /*
            The lock is created using the mtx variable
            It will wait until getting a wakeup signal from the condition variable (cv)
        */
        unique_lock<std::mutex> lck(mtx);
        cv.wait(lck, []{return ioReady;});
        // Set the ioReady to false. Prevents competing i/o in case of spurrious wakeup
        ioReady = false;
        cout << "Producer starting on core " << sched_getcpu() << endl;
        // Set the ioReady variable to true to set it up for the next waiting thread that gets to run
        ioReady = true;
        // Send a wakeup signal to one of the threads waiting using the condition variable (cv)
        cv.notify_one();
    }
 
    for (int i = 1; i <= NUMITEMS; i++)
    {
        // Obtain a lock based on the mtx variable.
        unique_lock<std::mutex> lck(mtx);
        // Wait until the condition variable sends a signal
        /*
            Notes for self:
            From https://www.cplusplus.com/reference/condition_variable/condition_variable/wait/

            If pred is specified (2), the function only blocks if pred returns false, 
            and notifications can only unblock the thread when it becomes true (which 
            is specially useful to check against spurious wake-up calls). 
            This version (2) behaves as if implemented as:
                    while (!pred()) wait(lck);
        */
        cv.wait(lck, []{return addToQueue;});

        cout << "Producer added " << i << " (qsz: " << aQueue->size() << ")" << endl;     
        // Add to the queue   
        aQueue->add(i);

        // If the queue has at least 1 item in it, set removeFromQueue to true
        if (aQueue->size() == 1)
        {
            removeFromQueue = true;
        }
        // If the queue is full, set addToQueue to false
        else if (aQueue->size() == BUFSIZE)
        {
            addToQueue = false;
        }
        // Send a wakeup signal to one of the threads waiting using the condition variable (cv)
        cv.notify_one();
    }

    {
        // Obtain a lock based on the mtx variable.
        unique_lock<std::mutex> lck(mtx);
        cv.wait(lck, []{return ioReady;});
        // Set the ioReady to false. Prevents competing i/o in case of spurrious wakeup
        ioReady = false;
        cout << "Producer ending" << endl;
        // Set the ioReady variable to true to set it up for the next waiting thread that gets to run
        ioReady = true;
        // Send a wakeup signal to one of the threads waiting using the condition variable (cv)
        cv.notify_one();
    }
    
}

void consumer()
{
    // Uses empty curly braces so that the unique lock will
    // die as it goes out of scope.
    // Lock the mutex so that the 
    {
        // Obtain a lock based on the mtx variable.
        unique_lock<std::mutex> lck(mtx);
        cv.wait(lck, []{return ioReady;});
        // Set the ioReady to false. Prevents competing i/o in case of spurrious wakeup
        ioReady = false;
        cout << "Consumer starting on core " << sched_getcpu() << endl;
        // Set the ioReady variable to true to set it up for the next waiting thread that gets to run
        ioReady = true;
        // Send a wakeup signal to one of the threads waiting using the condition variable (cv)
        cv.notify_one();
    }

    for (int i = 1; i <= NUMITEMS; i++)
    {
        // Obtain a lock based on the mtx variable.
        unique_lock<std::mutex> lck(mtx);

        // wait for the condition variable to send a wakeup signal to continue
        // Secondary check: removeFromQueue must be true
        cv.wait(lck, []{return removeFromQueue;});

        // Output mesage and remove from queue at the same time
        cout << "Consumer rem'd " << aQueue->remove() << " (qsz: " << aQueue->size() << ")" << endl;

        // If queue is empty set removeFromQueue to false
        if (aQueue->size() == 0)
        {
            removeFromQueue = false;
        }
        // If queue has at least one remaining space in it, set addToQueue to true.
        else if (aQueue->size() == (BUFSIZE - 1))
        {
            addToQueue = true;
        }
        // Send a wakeup signal to one of the threads waiting using the condition variable (cv)
        cv.notify_one();
    }

    {
        unique_lock<std::mutex> lck(mtx);
        cv.wait(lck, []{return ioReady;});
        // Set the ioReady to false. Prevents competing i/o in case of spurrious wakeup
        ioReady = false;
        cout << "Consumer ending" << endl;
        // Set the ioReady variable to true to set it up for the next waiting thread that gets to run
        ioReady = true;
        // Send a wakeup signal to one of the threads waiting using the condition variable (cv)
        cv.notify_one();
    }
}

File: cs415-Parallel-Programming/cpp-java-consumer-producer/ProdCons1.java
/*
    Brooke Czerwinski
    CS 415 Paralell Programming
    ProdCons1.java
    HW 1 

    Java-based producer/consumer program for parallel threads using a queue
*/

import java.util.LinkedList;
import java.util.Queue;

class ProdCons1{  
    // Required global parameters
    static int      BUFSIZE = 20;    // queue capacity
    static int      NUMITEMS = 100;  // total number of data items

    // Other global parameters
    static Object   queueAccess;                // Variable used for signalling/waiting
    static boolean  producerFinished = false;   // Whether or not the producer has finished adding to queue
    static boolean  queueFull = false;          // Whether or not the queue is full
    static boolean  queueEmpty = true;          // Whether or not the queue is empty
    static Queue<Integer> queue;                // Queue to be used throughout the program

    static Runnable producer = new Runnable()
    {
        public void run() {
            System.out.println("Producer starting");

            for (int i = 1; i <= NUMITEMS; i++)
            {
                synchronized(queueAccess)
                {
                    while (queueFull == true)
                    {
                        try{ queueAccess.wait(); }
                        catch(Exception e) { System.err.println("queueAccess.wait() = " + e.getMessage()); }
                    }

                    queue.add(i);
                    System.out.println("Producer added " + i + " (qsz: " + queue.size() + ")");

                    queueEmpty = false;
                    if (queue.size() >= BUFSIZE)
                    {
                        queueFull = true;
                    }

                    try { queueAccess.notify(); }
                    catch(Exception e) { System.err.println("queueAccess.notify() = " + e.getMessage()); }
                }
            }

            // Set producerFinished to true
            // then, in case there is a sleeping consumer thread 
            synchronized(queueAccess)
            {
                producerFinished = true;
                try { queueAccess.notify(); }
                catch(Exception e) { System.err.println("queueAccess.notify() = " + e.getMessage()); }
            }
            
            System.out.println("Producer ending");
        }
    };

    static Runnable consumer = new Runnable()
    {
        public void run() {
            boolean done = false;   // Variable to track whether or not we need to keep consuming

            System.out.println("Consumer starting");
            
            // Will keep the while loop going until there is no more to consume from the queue
            while (done != true)
            {
                // Establish that the following section requires atomic access
                synchronized(queueAccess)
                {
                    // We want to stop trying only when the producer is finished and the queue is empty
                    /*
                        ((producerFinished == true) && (queueEmpty == true)) 
                        negated = ((producerFinished != true) || (queueEmpty != true))
                    */
                    if ((producerFinished != true) || (queueEmpty != true))
                    {
                        // If the queue is empty we want to wait for a wakeup signal
                        /*
                            We will be woken up on that wait variable, but we need to make sure it's really for us.
                                For example, it could be another consumer thread signalling the producer that it has removed something
                                rather than the producer notifying us. So check for suprrious wakeup using while loop.
                            If it's not our turn we'll give up the lock so another blocked thread can use the signal
                        */
                        while (queueEmpty == true)
                        {  
                            // Wait for wakeup signal to reaquire access to the section
                            try { queueAccess.wait(); }
                            catch(Exception e) { System.err.println("queueAccess.wait() = " + e.getMessage()); }
                        }

                        // Remove item from the queue and print out to the user
                        System.out.println("Consumer rem'd " + queue.remove() + " (qsz: " + queue.size() + ")");
                        // Set queueFull to false since we know we have now taken at least one object out of it
                        queueFull = false;
                        // Check to see if that was the last item in the queue. Set queueEmpty to true, if so.
                        if (queue.size() == 0)
                        {
                            queueEmpty = true;
                        }

                        // Notify a thread that it can unlock
                        try { queueAccess.notify(); }
                        catch(Exception e) { System.err.println("queueAccess.notify() = " + e.getMessage()); }
                    }
                    else 
                    {
                        done = true; 
                    }
                }
            }
            
            System.out.println("Consumer ending");
        }
    };



    public static void main(String args[])  
    {
        // Initialize a queue
        queue = new LinkedList<>();
        queueAccess = new Object();

        // Initialize two threads to run producer() and consumer()
        Thread producerThread = new Thread(producer);
        Thread consumerThread = new Thread(consumer);
        
        try{
            // Start the two threads
            consumerThread.start();     // start consumer
            producerThread.start();     // start producer

            // Wait for the two threads to join back
            producerThread.join();      // Wait for producer to finish
            consumerThread.join();      // Wait for consumer to finish
        } catch (Exception e) { System.err.println(e.getMessage()); }

        System.out.println("Main: all done!");
    }
}